File: src/RCHttp.m
//
//  RCHttp.m
//  IMAGIN
//
//  Created by Baluta Cristian on 5/9/10.
//  Copyright 2010 ralcr. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "RCHttp.h"

@interface RCHttp () <NSURLSessionDelegate> {
	
	NSURL *_URL;
	NSURLSessionDataTask *_task;
}

@end

//static int _activeRequests = 0;

@implementation RCHttp

- (instancetype)initWithUrl:(NSString *)url {
	
	if (self = [super init]) {
		_URL = [NSURL URLWithString:url];
	}
	return self;
}

- (instancetype)initWithBaseUrl:(NSString *)url endpoint:(NSString *)endpoint {
	
	if (self = [super init]) {
		_URL = [NSURL URLWithString:[NSString stringWithFormat:@"%@%@%@", url, endpoint ? @"/" : @"", endpoint ? endpoint : @""]];
	}
	return self;
}

- (void)post:(NSDictionary *)dictionary
  completion:(void (^)(NSDictionary *))completionBlock
	   error:(void (^)())errorBlock {
	
	// Create POST variables
	NSMutableString *postStr = [[NSMutableString alloc] init];
	
	for (id key in dictionary) {
		[postStr appendFormat:@"%@=%@&", key, [dictionary objectForKey:key]];
	}
	
    NSData *postData = [postStr dataUsingEncoding:NSASCIIStringEncoding allowLossyConversion:YES];
    NSString *postLength = [NSString stringWithFormat:@"%lu", (unsigned long)[postData length]];
	
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setURL:_URL];
    [request setHTTPMethod:@"POST"];
    [request setValue:postLength forHTTPHeaderField:@"Content-Length"];
    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    [request setHTTPBody:postData];
	
	_task = [[NSURLSession sharedSession] dataTaskWithRequest:request
											completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
		if (error) {
			errorBlock();
		}
		else {
			NSError *err2 = nil;
			NSDictionary *responseDict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&err2];
			if (err2) {
				completionBlock(@{@"text":@"Json parse error"});
			}
			else {
				completionBlock(responseDict);
			}
		}
		[self downloadEnded];
	}];
	[_task resume];
	[self downloadStarted];
}

- (void)upload:(NSData *)data
	  withName:(NSString *)filename
	completion:(void (^)(NSDictionary *))completionBlock
		 error:(void (^)())errorBlock {

	// Setting up the request object now
	NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
	[request setURL:_URL];
	[request setHTTPMethod:@"POST"];
	
	// We always need a boundary when we post a file
//	NSString *boundary = @"---------------------------14737809831466499882746641449";
//	NSString *contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@", boundary];
//	[request addValue:contentType forHTTPHeaderField: @"Content-Type"];
//	
//	// Now lets create the body of the post
//	NSString *bound1 = [NSString stringWithFormat:@"\r\n--%@\r\n", boundary];
//	NSString *bound2 = [NSString stringWithFormat:@"\r\n--%@--\r\n", boundary];
//	NSString *contentDisposition = [NSString stringWithFormat:@"Content-Disposition: form-data; name=\"uploadFile\"; filename=\"%@\"\r\n", filename];
//	
//	NSMutableData *body = [NSMutableData data];
//	[body appendData:[bound1 dataUsingEncoding:NSUTF8StringEncoding]];
//	[body appendData:[contentDisposition dataUsingEncoding:NSUTF8StringEncoding]];
//	[body appendData:[@"Content-Type: application/octet-stream\r\n\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
//	[body appendData:[NSData dataWithData:data]];
//	[body appendData:[bound2 dataUsingEncoding:NSUTF8StringEncoding]];
//	
//	[request setHTTPBody:body];
	
	_task = [[NSURLSession sharedSession] uploadTaskWithRequest:request
													   fromData:data
											  completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
		
												  if (!error) {
													  NSError *jerror = nil;
													  NSDictionary *response_dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&jerror];
													  if (jerror) {
														  completionBlock(@{@"text":@"Json parse error"});
													  }
													  else {
														  completionBlock(response_dict);
													  }
												  }
												  else {
													  errorBlock();
												  }
												  [self downloadEnded];
											  }];
	[_task resume];
	[self downloadStarted];
}

- (void)cancel {
	[_task cancel];
}


#pragma mark Network indicator

- (void)downloadStarted {
	[UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
}

- (void)downloadEnded {
	[[NSURLSession sharedSession] getTasksWithCompletionHandler:^(NSArray *dataTasks, NSArray *uploadTasks, NSArray *downloadTasks) {
		if (dataTasks.count == 0 && uploadTasks.count == 0 && downloadTasks.count == 0) {
			[UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
		}
	}];
}

@end

File: src/RCHttp.swift
//
//  RCHttp.swift
//  Photography Places
//
//  Created by Baluta Cristian on 08/08/2014.
//  Copyright (c) 2014 Baluta Cristian. All rights reserved.
//

import UIKit

class RCHttp: NSObject {
	
	var url :String?;
	var task :NSURLSessionTask?;
	
	convenience init (_ baseURL:NSString, endpoint:NSString?) {
		self.init()
		let separator = endpoint != nil ? "/" : ""
		let end = endpoint != nil ? endpoint : ""
		url = String("\(baseURL)\(separator)\(end!)")
	}
	
	
	//pragma mark post data sync and async

	func post (dictionary:Dictionary<String,AnyObject>, completion:NSDictionary -> Void, errorHandler:NSDictionary -> Void) {
	
		var postStr :NSString = "";
		for (key, vale) in dictionary {
			postStr = "\(postStr)\(key)=\(vale)&";
		}
		
		let postData :NSData = postStr.dataUsingEncoding(NSASCIIStringEncoding, allowLossyConversion:true)!;
		let postLength = String (postData.length);
		
		let request = NSMutableURLRequest()
		request.URL = NSURL(string:url!)
		request.HTTPMethod = "POST"
		request.setValue(postLength, forHTTPHeaderField:"Content-Length")
		request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField:"Content-Type")
		request.HTTPBody = postData;
		
		let session = NSURLSession( configuration: NSURLSessionConfiguration.ephemeralSessionConfiguration());
		task = session.dataTaskWithRequest(request, completionHandler: {(data, response, error) in
			
			// notice that I can omit the types of data, response and error
			
//			RCLogO( NSString(data:data, encoding: 0));

			if (error == nil) {
				var json: NSDictionary = ["text":"Json parse error"]
				if let d = try? NSJSONSerialization.JSONObjectWithData(data!, options:NSJSONReadingOptions.AllowFragments) {
					json = d as! NSDictionary
				}
				completion(json)
			}
			else {
				errorHandler(["text":"Download error"])
			}
		})
		task?.resume()
	}
	
	func upload (data:NSData, filename:String, completion:NSDictionary -> Void, error:NSDictionary -> Void) {
		
		let request :NSMutableURLRequest = NSMutableURLRequest();
		request.URL = NSURL(string:url!);
		request.HTTPMethod = "POST";
//		request.setValue(postLength, forHTTPHeaderField:"Content-Length");
		request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField:"Content-Type");
//		request.HTTPBody = postData;
		
		let session = NSURLSession( configuration: NSURLSessionConfiguration.ephemeralSessionConfiguration());
		task = session.uploadTaskWithRequest(request, fromData: data, completionHandler: { (data_, response_, error_) -> Void in
			var json :NSDictionary? = nil;
			
			if (error_ != nil) {
				let e: NSErrorPointer? = nil;
				var response_dict :AnyObject?
				do {
					response_dict = try NSJSONSerialization.JSONObjectWithData(data_!, options:NSJSONReadingOptions.AllowFragments)
				} catch let error as NSError {
					e!.memory = error
					response_dict = nil
				} catch {
					fatalError()
				};
				//				println(response_dict);
				if (e! == nil) {
					json = ["text":"Json parse error"];
				}
				else {
					json = response_dict as? NSDictionary;
				}
				completion(json!);
			}
			error(["text":"Dwnlad error"]);
		})
		task?.resume()
	}
	
	func cancel () {
		task?.cancel()
	}
	
	func downloadStarted () {
		UIApplication.sharedApplication().networkActivityIndicatorVisible = true;
	}
	
	func downloadEnded () {
		UIApplication.sharedApplication().networkActivityIndicatorVisible = false;
	}
}
